rfd/00xx-snowflake-support.md

```diff
@@ -0,0 +1,171 @@
+    Teleport->>client: Forward the response
+```
+
+With this approach the real session token never leaves the Teleport and any client is not able to bypass Teleport
+by extracting the session token.
+Snowflake session token needs to be refreshed every X minutes. For that purpose Teleport would have to internally refresh
+the token as client doesn't have access to the token (login response contains the time for how long the token is valid).
+
+#### Query logging
+
+Queries can be extracted from requests sent to `/queries/v1/query-request` endpoint, field `data.sqltext`.
+
+#### User authentication
+
+Username can be extracted from `/session/v1/login-request` requests, field `data.LOGIN_NAME`. The extracted username
+can be then use to authenticate the user against the Teleport RBAC.
+
+### CLI support 
+
+`snowsql` is the official Snowflake CLI tool, and it doesn't support TLS terminated proxies as described 
+in https://docs.snowflake.com/en/user-guide/snowsql-start.html#using-a-proxy-server, and it uses plain text communication 
+when connecting to the Snowflake on a different port that 443. 
+
+`snowsql` requires to use one of the available authentication method. By default, it asks for password unless other 
+authentication option is specified. The other alternatives are SSO and JWT/Private key (https://docs.snowflake.com/en/user-guide/snowsql-start.html).
+Teleport could use JWT/Private key for authentication. In this flow `tsh` would export user's private in PKCS8 format (at that's the only one supported by `snowsql`),
+and `snowsql` could use it to sign the JWT token.
+
+I think there are two ways on how we can support the `snowsql`: 
+
+1. `snowcli` can connect to a custom host **without** TLS on any port different from 443. 
+   Because we cannot start `tsh proxy db` on port 443 (all systems require the root account for that) we could 
+   start `tsh` proxy on a different port as `tsh` can encrypt the transmission to the Teleport proxy. This approach of course
+   has one big drawback; the connection is not end-to-end encrypted as communication between `tsh` and `snowsql` uses plain text.
+2. Despite the documentation I was able to use TLS terminated proxy with `tsh`. In this flow `snowsql` sends the traffic 
+   to Teleport by using HTTP proxy protocol. Teleport then was able to terminate the TLS connection and read the traffic.
+   In order to make it to work I exported the Teleport issued CA as a `CURL_CA_BUNDLE` environment variable (this is undocumented feature).
+
+Below example shows how to start `snowsql` in that mode:
+
+```shell
+export https_proxy=localhost:2000
+export http_proxy=localhost:2000
+export CURL_CA_BUNDLE=/opt/homebrew/etc/ca-certificates/cert.pem # includes the Teleport issued CA
+export REQUESTS_CA_BUNDLE=/opt/homebrew/etc/ca-certificates/cert.pem
+
+snowsql --noup -a im56867.us-east-2.aws -u teleport-user --private-key-path ${HOME}/.tsh/keys/example.com/bob.pkcs8 -h me.localhost
```

Comment:
Does the snowsql  tools trigger some kind of  HTTP requests that are not stricte related to Snowflake SQL API. For example client upload version check ?. I that case I think that is is good idea to forward the request to the url without `local_proxy->teleport_proxy->app agent` flow.

Reply:
We could use `NO_PROXY` environment variable, https://docs.snowflake.com/en/user-guide/snowsql-start.html#using-a-proxy-server that that would require to list all domains that should skip the proxy.